map-reduce/word-count.c

```diff
@@ -4,70 +4,103 @@
+#define HLIST_HEAD(name) struct hlist_head name = {  .first = NULL }
+#define INIT_HLIST_HEAD(ptr) ((ptr)->first = NULL)
 
-static inline struct list_entry *list_next(struct list_entry *root,
-                                           struct list_entry *current)
+static inline void INIT_HLIST_NODE(struct hlist_node *h)
 {
-    if ((root == root->next) || (current->next == root)) return NULL;
-    return current->next;
+	h->next = NULL;
+	h->pprev = NULL;
 }
 
-/* FIXME: this forbids having 2 list_for_each in the same function, because the
- * variable __ptr will be defined twice, which results in a compilation error.
- * The __ptr is necessary because some functions delete iter while traversing
- * the list.
- */
-#define list_for_each_forward(root_ptr, iter)                                 \
-    struct list_entry *__ptr;                                                 \
-    for (iter = (root_ptr)->next, __ptr = (struct list_entry *) (iter)->next; \
-         iter != (root_ptr); iter = (typeof((iter))) __ptr,                   \
-        __ptr = (struct list_entry *) iter->next)
-
-#define list_for_each(root_ptr, iter) list_for_each_forward(root_ptr, iter)
+static inline int hlist_empty(const struct hlist_head *h)
+{
+	return !h->first;
+}
 
-static inline void list_root_init(struct list_entry *root)
+static inline void hlist_add_head(struct hlist_node *n, struct hlist_head *h)
 {
-    root->next = root->prev = root;
+	struct hlist_node *first = h->first;
+	n->next = first;
+	if (first)
+	    first->pprev = &n->next;
+	h->first = n;
+	n->pprev = &h->first;
 }
 
-static inline void list_add(struct list_entry *root, struct list_entry *entry)
+#include <stdbool.h>
+
+static inline bool hlist_is_singular_node(struct hlist_node *n, struct hlist_head *h)
 {
-    struct list_entry *prev_entry = root;
-    struct list_entry *next_entry = root->next;
-    entry->next = next_entry, entry->prev = prev_entry;
-    prev_entry->next = entry, next_entry->prev = entry;
+	return !n->next && n->pprev == &h->first;
 }
 
-#define list_add_prev(root, entry) list_add((root)->prev, (entry))
+#define container_of(list_ptr, container_type, member_name)               \
+    ({                                                                    \
+        const typeof(((container_type *) 0)->member_name) *__member_ptr = \
+            (list_ptr);                                                   \
+        (container_type *) ((char *) __member_ptr -                       \
+                            offsetof(container_type, member_name));       \
+    })
 
-#define list_empty(root) (root == (root)->next)
+#define hlist_entry(ptr, type, member) container_of(ptr,type,member)
 
-#include <stdint.h>
-#include <stdlib.h>
+#define hlist_first_entry(head, type, member) \
+    hlist_entry((head)->first, type, member) 
+
+#define hlist_for_each(pos, head) \
+	for (pos = (head)->first; pos ; pos = pos->next)
+
+#define hlist_for_each_safe(pos, n, head) \
```

Comment:
The macro `hlist_for_each_safe` is unused. Remove.

Reply:
Done!